terraform/node_resource_abstract_instance.go

```diff
@@ -2084,18 +2083,12 @@ func (n *NodeAbstractResourceInstance) apply(
 		// deleted then our next refresh will detect that and fix it up.
 		// If change.Action is Create then change.Before will also be null,
 		// which is fine.
-		newVal = change.Before
-
-		// If we're recovering the previous state, we also want to restore the
-		// the tainted status of the object.
-		if state.Status == states.ObjectTainted {
-			newStatus = states.ObjectTainted
-		}
+		newState = state.DeepCopy()
 	}
 
 	if !newVal.IsNull() { // null value indicates that the object is deleted, so we won't set a new state in that case
 		newState = &states.ResourceInstanceObject{
-			Status:              newStatus,
+			Status:              states.ObjectReady,
```

Comment:
What happens if the object was tainted? Should this be `newState.Status`?

Reply:
James pointed out what I was missing in slack:
> newStatus is only set to tainted if the the value is null, but then only assigned if the value is not null

Reply:
`newStatus` could only be set to `ObjectTainted` is `newVal` is null, but this requires that `newVal` is not null so it would never be set; so there is no change in behavior here.

I do think that we may lose the `ObjectTainted` status if a provider errors when removing a tainted object _and_ returns a non-null value. Most providers wouldn't be capable of doing this right now, but I'll make sure it's covered in another PR.